cpp/include/cudf/ast/detail/expression_evaluator.cuh

```diff
@@ -461,15 +527,27 @@ struct expression_evaluator {
           plan.data_references[plan.operator_source_indices[operator_source_index++]];
         auto input_row_index =
           input.table_source == table_reference::LEFT ? left_row_index : right_row_index;
-        type_dispatcher(input.data_type,
-                        *this,
-                        output_object,
-                        input_row_index,
-                        input,
-                        output,
-                        output_row_index,
-                        op,
-                        thread_intermediate_storage);
+        if constexpr (has_complex_type) {
+          type_dispatcher(input.data_type,
+                          *this,
+                          output_object,
+                          input_row_index,
+                          input,
+                          output,
+                          output_row_index,
+                          op,
+                          thread_intermediate_storage);
+        } else {
+          type_dispatcher<id_to_primitive_type>(input.data_type,
+                                                *this,
+                                                output_object,
+                                                input_row_index,
+                                                input,
+                                                output,
+                                                output_row_index,
+                                                op,
+                                                thread_intermediate_storage);
```

Comment:
The parameters in two calls to `type_dispatcher` are identical. Since `has_complex_type` is already `constexpr`, we only need the second call, which covers both situations.